src/commands/login.rs

```diff
@@ -0,0 +1,400 @@
+                    serde_json::to_string_pretty(
+                        &create_device_code(&Client::new(connection_config)).await?
+                    )?
+                );
+                return Ok(());
+            }
+
+            let token: TokenInfo;
+            if let Some(device_code) = self.check_device_code {
+                let client = Client::new(connection_config);
+                match client.login(device_code).await {
+                    Ok(token_info) => {
+                        if token_info.token.is_some() {
+                            println!("{}", serde_json::to_string_pretty(&token_info)?);
+                            token = token_info;
+                        } else {
+                            println!(
+                                "{}",
+                                serde_json::to_string_pretty(&json!({ "status": "waiting" }))?
+                            );
+                            return Ok(());
+                        }
+                    }
+                    Err(e) => {
+                        return Err(e);
+                    }
+                };
+            } else {
+                token = github_token(connection_config).await?;
+            }
+
+            login_connection = LoginConnection {
+                url,
+                danger_accept_invalid_certs: self.insecure,
+                token: token.token.unwrap_or_default(),
+                expiration: token.expiration.unwrap_or_default(),
+                bindle_url: None,
+                bindle_username: None,
+                bindle_password: None,
+            };
+        }
+
+        std::fs::write(path, serde_json::to_string_pretty(&login_connection)?)?;
+        Ok(())
+    }
+}
+
+async fn github_token(
+    connection_config: ConnectionConfig,
+) -> Result<cloud_openapi::models::TokenInfo> {
+    let client = Client::new(connection_config);
+
+    // Generate a device code and a user code to activate it with
+    let device_code = create_device_code(&client).await?;
+
+    println!(
+        "Open {} in your browser",
+        device_code.verification_url.clone().unwrap(),
+    );
+
+    println!(
+        "! Copy your one-time code: {}",
+        device_code.user_code.clone().unwrap(),
+    );
+
+    // The OAuth library should theoretically handle waiting for the device to be authorized, but
+    // testing revealed that it doesn't work. So we manually poll every 10 seconds for fifteen minutes.
+    const POLL_INTERVAL_SECS: u64 = 10;
+    let mut seconds_elapsed = 0;
+    let timeout_seconds = 15 * 60;
+
+    // Loop while waiting for the device code to be authorized by the user
+    loop {
+        if seconds_elapsed > timeout_seconds {
+            bail!("Timed out waiting to authorize the device. Please execute `spin login` again and authorize the device with GitHub.");
+        }
+
+        match client.login(device_code.device_code.clone().unwrap()).await {
+            Ok(response) => {
```

Comment:
This needs to check whether `response.token` is `None` and continue to wait if so.

Reply:
fixed